src/core/qgsdatasourceuri.cpp

```diff
@@ -53,6 +53,7 @@ QgsDataSourceUri::QgsDataSourceUri( const QString &u )
       i++;
 
     const QString pname = uri.mid( start, i - start );
+    mParamKeys.insert( pname );
```

Comment:
I think this needs handling in more places -- e.g. QgsDataSourceUri::setParam, QgsDataSourceUri::removeParam, QgsDataSourceUri::setEncodedUri, and all the other set... methods too. 

In fact, I think it'd be cleaner to remove mParamKeys and build this list dynamically in QgsDataSourceUri::parameterKeys(), by starting with mParams.keys() and adding keys for the dedicated members like schema, database, etc (when those are non-empty).

Reply:
done